### DISCUSSION TITLE
Simulation Results Refactor

### DISCUSSION THREAD

**Discussion by rnugent3** [HydrologicEngineeringCenter/fda-model#88] [General]
_We need to refactor the code that gives us simulation results. Alternative and alternative comparison report results are a separate beast._

**Problems include:**
1. We don't have a good results interface. The current interface is empty. 
2. Accessing results is messy. For example, getting the mean aep happens like so: 
`double actual = results.Thresholds.ThresholdsDictionary[thresholdID].Performance.MeanAEP();`
3. Not all needed results are available. For example, the aep and ead histograms are private. 

**Opportunities include:**
1. Simplify code, make accessing results more intuitive 
2. Design to an interface 

**Objectives include:** 
1. Get all of the results we need 
2. Get the results we need easily 

**Constraints include:**
1. ?

**What are the simulation results we need:**
1. Access to EAD histogram (display histogram and summary statistics) by damage category and aggregate over damage categories 
2. Access to AEP histogram (display histogram and summary statistics) for each threshold
3. CNEP for a given set of standard frequencies for each threshold
4. LTEP for a given set of periods for each threshold.

**Comment by rnugent3** [comment]
17 Dec 21 Live Discussion between @HenryGeorgist @rnugent3 

**Consider: what would I like the path to the result of interest to be? Refactor to make this better.**
Working off the example above, the facade pattern:
results.MeanAEP(ThresholdID) and do the work above in a closet, close door. 
Then, MeanAEP goes into the interface. 

Two interfaces:
ISimulationResults
ISimulationPreviewResults (damage frequency and ead)

**Reply by HenryGeorgist** [reply]
Maybe IPreviewSimulationResults?

**Reply by HenryGeorgist** [reply]
```
public interface ISimulationResults{
     public statistics.Idistribution AEPHistogram();
     public statistics.Idistribution EADHistogram(thresholdID);
     public double ConditionalNonExceedenceProbability(conditionalProbability);
     public double LTEP(thresholdID, years);
     public double MeanAEP();
     public double MeanEAD(thresholdid);

**Reply by HenryGeorgist** [reply]
what about number of iterations? if it was converged? are there other factors we care about for a simulation that may not map to ER 1105-2-101?

**Comment by rnugent3** [comment]
on the interface, we'll need a SetDamageFunction and GetDamageFunction. We'll need a signal in the code that we have a preview. e.g.

if results object is a IPreviewResults object, return a boolean or something: if (resultsObject is IPreviewResults)

**Reply by HenryGeorgist** [reply]
i think this only applies on the IPreviewResults object, because we arent intending on storing the damage frequency by damage category for each iteration.
